syntax = "proto3";

package ai4eu.orchestrator;

message OrchestrationConfiguration {
  string blueprint = 1;
  string dockerinfo = 2;
  repeated string protofiles = 3;
}

message OrchestrationStatus {
  string message = 1;
  bool success = 2;
  int32 code = 3;
  int32 active_threads = 4;
}

message RunLabel {
  string label = 1;
}

message OrchestrationObservationConfiguration {
  string name_regex = 1;
  string component_regex = 2;
}

message OrchestrationEvent {
  int32 run = 1;
  string name = 2;
  string component = 3;
  string detail = 4;
}

service Orchestrator {
    rpc get_status (RunLabel) returns (OrchestrationStatus);
    rpc initialize (OrchestrationConfiguration) returns (OrchestrationStatus);
    rpc run (RunLabel) returns (OrchestrationStatus);
    rpc observe (OrchestrationObservationConfiguration) returns (stream OrchestrationEvent);
}
